clang/lib/Frontend/TextDiagnostic.cpp

```diff
@@ -1288,11 +1291,33 @@ void TextDiagnostic::emitSnippet(StringRef SourceLine,
 
   // Print the source line one character at a time.
   bool PrintReversed = false;
+  bool HighlightingEnabled = DiagOpts->ShowColors;
   size_t I = 0;
   while (I < SourceLine.size()) {
     auto [Str, WasPrintable] =
         printableTextForNextCharacter(SourceLine, &I, DiagOpts->TabStop);
 
+    // Just stop highlighting anything for this line if we found a non-printable
+    // character.
```

Comment:
Instead, can we switch to a different text color for the cases where we print a non-printable character as `<U+1234>`? Eg, black text on red background.

Reply:
IIRC my problem was that the indices we save in the `StyleRange` don't match anymore once we wrote `<U+1234>`.